### DISCUSSION TITLE
Integrating Airflow with SLURM

### DISCUSSION THREAD

**Discussion by ecodina** [apache/airflow#24076] [Q&A] [answered]
Hello,

I am currently trying to use Airflow in an HPC which uses Slurm to distribute jobs over different nodes. We don't wish to replace Slurm but use Airflow to launch jobs.

I have tried to integrate both systems during the last two weeks, but haven't found a proper way to do it. I haven't found information on any site either.

**Approach 1:** create a custom Executor

In this case, the custom executor generates the Slurm command: `sbatch [options] airflow tasks run dag_id task_id run_id`.
The executor then regularly checks the `squeue` command to find when the job has finished.

I found some problems:

- The command `airflow tasks run dag_id task_id run_id` always returns a successful exit code. Therefore, Slurm considers that all jobs have finished properly. For the sake of data correctness, we would prefer that both systems showed the real state of the job. Is there any way for `airflow tasks run` to propagate any task error? The `-N` flag didn't help.
- Surprisingly, when the task failed (the BashOperator raised an AirflowException), it automatically flagged the task as Failed and stayed this way even if I flagged the task as OK from the executor (since Slurm reported that the job had finished successfully).

**Approach 2:** create a custom Operator (based on the BashOperator)

_a) Using srun_
In this case, the custom operator generates the Slurm command: `srun [options] /path/to/script/to/run.sh`.
Since `srun` is interactive, all the job output can be easily captured into the Airflow task log.

However:
- If the Airflow scheduler fails for any reason, the whole job gets canceled. Since we are running some time-consuming (10+ hours) computations, this can seriously disrupt the business.

_b) Using sbatch_
In this case the custom operator generates the Slurm command:  `sbatch [options] /path/to/script/to/run.sh`.
The operator then continuously monitors the Slurm queue and raises an AirflowException if it fails, so the task shows as Failed in the UI.

However:
- If Airflow fails for any reason, the job continues running (since Slurm is responsible for the execution). However, if we restart the Airflow scheduler, it loses track of that job (it shows as failed), which means that we would need to manually check whether the jobs have finished correctly or not.

Does anyone have experience with this issue? I would appreciate your input!

Thanks :)

**Comment by potiuk** [comment]
No idea about SLURM, but:

Approach 1:

You need to implement your executor that it will also monitor and report the task status back. Since Airlfow is a distributed system, simple "task error code" is not enough to keep the sate. The task might fail for various reasons, it can then be retried when - for example - celery worker is restarted and there are various edge cases. There are two ways state might be updated - tasks can modify their own state when they succeed in the database or the monitoring (executor) determines that the tasks are failed or did not have time to update the state and updates it for the task. This is due to distributed nature of Aiflow, potential failover scenarios and the like - so sime "exit code" is not good indicator and you should not base task state on that. Writing your executor means implementing a lot of failover and failure edge cases and it's quite a complex task.

Approach 2:
* Airflow operator MUST be idempotent. This is an absolute prerequisite - and when yoy develop your operator you have to make sure to implement idempotency (idempotency in the sense that either reruning the task overrides the data generated by the previous run or that it checks that the run is either running or complete and skips running. This is the job of the operator.

**Comment by ecodina** [comment, marked as answer]
Hi @potiuk 

Thank you for your answer!

We've ended up developing a deferrable operator and a trigger. It first submits the job to Slurm and then it defers itself until the trigger detects a state change / new output from the slurm job's log file. Depending on the slurm's state, we defer the operator again, finish OK or raise an AirflowException.

Since triggers are able to run in a highly-available fashion, we will be able to to restart Airflow for any reason without losing track of the already submitted Slurm jobs.

As for idempotency, in our case rerunning the task overrides the data.

**Reply by potiuk** [reply]
Yep. This is one of the good siodes of the deferrable operator ! Thanks for pointing it out and glad that Deferrables find their way to mainstream use :)

**Reply by janvainer** [reply]
@ecodina very cool solution! Could you please share an example code for the deferrable operator? It would help me with essentially the same problem that you were solving! :)

**Reply by ecodina** [reply]
@janvainer Since the code is quite long, I've created a gist with the deferrable operator & trigger: https://gist.github.com/ecodina/157b5dc44b79b13fe296b1275b4f0967

Some comments are in Catalan, although it is mainly in English.

**Reply by janvainer** [reply]
This is awesome! Thank you very much :)

**Reply by janvainer** [reply]
@ecodina did you experiment with running python tasks using the @task airflow API? It seems that with slurm, one has to use shell to run separate tasks.

**Reply by ecodina** [reply]
@janvainer since this is an operator, based on the BashOperator, you can only run bash commands / scripts, so you can't run python tasks directly.

**Reply by jimklo** [reply]
@ecodina I have to say this is pretty slick.  I actually have a variation of this same problem.  My SLURM job is queued in a DockerOperator using fabric, however I wasn't sure how to deal with checking for completion.  I like the trigger solution.

**Reply by leozitor** [reply]
@ecodina did you do some improvements in this operator after? Looks pretty well build, I am trying to do some modifications to run the slurm commands over ssh as in my case I cannot use the hpc as worker directly, then i should run the jobs over ssh commands.

**Reply by ecodina** [reply]
@leozitor we may have done a few tweaks to adapt it to our use case but overall the code remains as is. I don't think you would need a lot of changes to connect via ssh, but the triggerer checks the state of the task and logger every very few seconds -> maybe you should keep a ssh session active all the time to avoid reconnecting?

**Reply by leozitor** [reply]
@ecodina I am still learning how airflow works under the hood, so I am still figuring out the idea of your code, but for now I did a quick solution to my case creating a custom sensor that ssh the HPC keeps verifying the state of the job and reschedules while is not completed, but sure not well designed as your operator ( the sync part still bugs my mind haha), later i will try to do some changes to  that operator to be able to use ssh, but for now the sensor is working and I can move on with the development.
If I end up improving it i will share here, thanks for the reply anyway!

**Reply by FeryET** [reply]
I would like to see if anyone has implemented this, it sounds intimidating.

**Reply by leozitor** [reply]
> I would like to see if anyone has implemented this, it sounds intimidating.

do you mean implement what exatcly? ecodina implemented a whole deferable operator to work with slurm, I just implemented a small SensorOperator to work with slurm through ssh.

**Reply by elehcim** [reply]
Thanks @ecodina for your gist, I really learned a lot! I have a similar use case to @leozitor, I'd like to use slurm through SSH as well. I managed to create an SSHSlurmOperator with a ssh hook using the [ssh hook provider](https://airflow.apache.org/docs/apache-airflow-providers-ssh/stable/_api/airflow/providers/ssh/hooks/ssh/index.html).
And for the trigger I used [asyncssh](https://asyncssh.readthedocs.io/en/latest/).

I share it here: https://gist.github.com/elehcim/52150ab6711d49b63d175bd4c8fe3efd

For the moment, each time the trigger issues commands on the remote cluster using asyncssh a new connection is created. It would be better to use just one connection and open new channels on it. Not sure how to do it. Please, if anybody has improvements, do not hesitate to fork it.
@leozitor , if you have an updated version of your sensor it would be nice if you can share it. Thank you very much!

**Reply by ecodina** [reply]
Good to hear, @elehcim! Another thing you could improve (and I have been working on that the last few weeks in my company) is to use a new Airflow feature. You can now log from the trigger, so there is no need for the operator -> trigger -> operator -> .... anymore. This should improve stability.

**Reply by elehcim** [reply]
Nice, @ecodina thanks. Could you point me to the new Airflow feature you mention? Thanks

**Reply by ecodina** [reply]
Sure, @elehcim . Just use "self.log.info" from inside the trigger and only yield a TriggerEvent when the slurm state is completed (failed, success, ...).

Since triggers are HA and may be moved around, you'll need to find a way to store how many log lines you've already read, in case the trigger is reinstanciated elsewhere. In my case, I use the "cleanup" method to store the # of lines read for a task in an external database. I then remove this entry on the operator method that gets called after the trigger.

Maybe in the future it will be possible to reserialize the trigger on-demand so this database won't be necessary, but it is not possible now AFAIK.

**Reply by bejota** [reply]
Thank you. This is a great solution. Has anyone considered SLURM's REST API rather than SSH?

**Reply by ecodina** [reply]
@bejota Since my last answer we have migrated the service to use a new architecture. Airflow now adds the task information (at the SlurmOperator) at a Redis database. When a daemon receives a new Redis object, it processes it and runs the job in the slurm cluster, as well as updates the state at the Redis database. The SlurmTrigger reads the job state in the Redis database and updates the Airflow task accordingly.

I have applied for a presentation in this year's Airflow Summit. If I'm accepted (hopefully!) I'll be able to present this solution in greater detail.

**Reply by bejota** [reply]
Very interesting, @ecodina. I would love to learn more. Good luck with the presentation!

**Reply by bhulsey** [reply]
After rebuilding SLURM with support for its REST API, I'm able to submit jobs using the `SimpleHttpOperator` and monitor them with the deferrable `HttpSensor`. I've only run a toy example but it was much easier for me than navigating SSH from an Airflow container.

**Reply by elehcim** [reply]
Hi @bhulsey , I find this approach very promising and simpler. Could you please share a toy example, even minimal, so that I can test and evaluate it? Cheers, thanks

**Reply by ecodina** [reply]
Hi @elehcim and @bhulsey. Before using Redis, we used to query directly the `sacct` command for each task, every 5 seconds in order to view the job state and update it in Airflow almost on real-time. Once we had over 400 DAGs running, we found out that the cluster started to go slow.

This is why we added the Redis layer. We now run a single `sacct` command, which queries all the jobs. and updates Redis. Then Airflow reads this database. By doing this, we found a massive improvement of the system's saturation.

I imagine that the Slurm API would be similar to running `sacct`, but you'll have to test it.

**Reply by bejota** [reply]
@ecodina I'm also concerned about the load on SLURM. I admit your solution sounds like the best in general. However, the 2 most likely scenarios we have are 

1.  <10 independent jobs
2.  a job array that is submitted/monitored from a single dag

@elehcim Below is a very toy example, with a few caveats.

1. I don't have nodes connected to the test cluster, so I'm using pings in place of submit/query
2. `HttpSensor` does not defer to the `triggerer` when using `response_check` parameter. The `trigger` only checks for a valid http response code. The `response_check` parameter, which I need to check job status, causes the task to stay in the scheduler. So, I will need to adapt `HttpTrigger` or `HttpSensorTrigger` to check job status. 

_Comment out `response_check` and use a bad endpoint to verify the task is deferred to `triggerer`._

```
import pendulum
from airflow import DAG
from airflow.providers.http.operators.http import SimpleHttpOperator
from airflow.providers.http.sensors.http import HttpSensor

dag = DAG(
    dag_id="slurm_http_dag",
    start_date=pendulum.datetime(2024, 5, 1, tz="UTC"),
    schedule=None,
    dagrun_timeout=pendulum.duration(minutes=60),
    catchup=False,
)


submit_slurm_job = SimpleHttpOperator(
    task_id="submit_slurm_job",
    http_conn_id="http-slurm",
    method="get",
    endpoint="slurm/v0.0.40/ping",
    log_response=True,
    # deferrable=True,
    dag=dag,
)


check_slurm_job = HttpSensor(
    task_id="check_slurm_job",
    http_conn_id="http-slurm",
    method="get",
    endpoint="slurm/v0.0.40/ping",
    # bad endpoint without response_check defers to triggerer
    # endpoint="slurm/v0.0.40/pings",
    # response_check=lambda response: "UP" in response.text, # pass response_check
    response_check=lambda response: "foo" in response.text,  # fail response_check
    poke_interval=10,
    timeout=60,
    deferrable=True,  # does not defer to triggerer if using response_check
    dag=dag,
    # mode="reschedule",
)

submit_slurm_job >> check_slurm_job

```

**Comment by xiuliren** [comment]
I found a DaskExecutor. Dask can use slurm to launch a cluster. I have never tried this idea though.

**Comment by AndreaRiboniSF** [comment]
@ecodina did you think of a way to "stream" the logs coming from the slurm job to airflow?

**Reply by ecodina** [reply]
Hello @AndreaRiboniSF 

We developed a SlurmOperator, SlurmTrigger and some deamons. You can have more information in my last year presentation at the Airflow Summit (https://youtu.be/ol6k7df3Kr0)

The communication between Airflow and the deamons is done via Redis, but this communication is limited to task parameters (code, resources needed and Slurm job status). 

In the cluster, all our jobs generate logs in the same folder, with a predefined name (`slurm-{jobid}.out`). We made this folder available in the virtual machine where Airflow runs (this will depend on your setup: on how the shared filesystem works), so that the trigger can access the log by just knowing the jobid.

Anlther alternative could be to stream the logs via Redis. For that, you'd probably need a 3rd daemon that should monitor all the files and send the lines. You could also see if using fluentd or the ELK stack could help you out.

**Reply by bejota** [reply]
Hello @ecodina ,
We have a similar workflow as you describe. We are collecting seismic data, so replace "Simulation" with "Migration". We still have individual job or job array monitoring in place but have not run at full scale yet, so not much to report there. 

We are still using Airflow version 2.10. Have you tried migrating to Airflow v3 yet? If so, have you made any changes in the workflow?

**Reply by ecodina** [reply]
We've been using this workflow for the last 3 - 4 years, generating 2.2M dagruns (10x that for task runs, probably) per year. We have not noticed any problems.

We are also in 2.10, but plan on upgrading to 3.1 in the near months. I believe we won't need changes in the workflow since I believe Operators and Triggers have not changed. If they have, it will probably be a minimal change.

**Reply by AndreaRiboniSF** [reply]
I ended up the similar file-based logging with the trigger. I implemented the system using Airflow 3.1.0, but currently it is in a local environment managed with docker compose
